File: Code/methods.py
import math

def generate_Ms(str):
    Ms = []
    str_index = 0
    for i in range(16):
        if(i == 14):
            Ms.append(0)
        elif(i == 15):
            Ms.append(8 * len(str))
        else:
            M = 0

            if(str_index < len(str)):
                for j in range(4):
                    if(str_index < len(str)):
                        M = M | (ord(str[str_index]) << (24 - 8*(str_index % 4)))
                        if(str_index + 1 == len(str)):
                            if((str_index + 1) % 4 == 0):
                                Ms.append(M)
                                M = 1 << 31
                            else:
                                M = M | (1 << (23 - 8*(str_index % 4)))

                    str_index += 1

            Ms.append(M)

    return Ms

def generate_Ks():
    Ks = []
    for i in range(64):
        Ks.append(int(abs(math.sin(i + 1))*pow(2,32)))

    return Ks

def rotate(num, shift):
    shift_portion = num >> (32 - shift)
    eraser = 0b11111111111111111111111111111111
    num &= eraser >> shift

    return (num << shift) | shift_portion

def F(A, B, C, D, M, K, S):
    result = (B & C) | ((~B) & D)

    result = (result + A) % pow(2, 32)

    result = (result + M) % pow(2, 32)

    result = (result + K) % pow(2, 32)

    result = rotate(result, S)

    result = (result + B) % pow(2, 32)

    return [result, B, C, D]

def G(A, B, C, D, M, K, S):
    result = (B & D) | (C & (~D))

    result = (result + A) % pow(2, 32)

    result = (result + M) % pow(2, 32)

    result = (result + K) % pow(2, 32)

    result = rotate(result, S)

    result = (result + B) % pow(2, 32)

    return [result, B, C, D]

def H(A, B, C, D, M, K, S):
    result = B ^ C ^ D

    result = (result + A) % pow(2, 32)

    result = (result + M) % pow(2, 32)

    result = (result + K) % pow(2, 32)

    result = rotate(result, S)

    result = (result + B) % pow(2, 32)

    return [result, B, C, D]

def I(A, B, C, D, M, K, S):
    result = C ^ (B | (~D))

    result = (result + A) % pow(2, 32)

    result = (result + M) % pow(2, 32)

    result = (result + K) % pow(2, 32)

    result = rotate(result, S)

    result = (result + B) % pow(2, 32)

    return [result, B, C, D]

File: Code/methods.c
#include <stdio.h>
#include "methods.h"

void generate_Ms(char *fileName, unsigned int *Ms) {
    int fd = open(fileName, O_RDONLY);
    if(fd == -1) {
        printf("%s\n", strerror(errno));
    }

    struct stat stat_buffer;
    stat(fileName, &stat_buffer);
    int str_len = stat_buffer.st_size;

    int str_index = 0;
    for(int i = 0; i < 16; i++) {
        if(i == 14) {
            Ms[i] = 0;
        }
        else if(i == 15) {
            Ms[i] = 8 * str_len;
        }
        else {
            unsigned int M = 0;

            if(str_index < str_len) {
                for(int j = 0; j < 4; j++) {
                    char chr = 0;
                    read(fd, &chr, sizeof(char));
                    if(str_index < str_len) {
                        M = M | (chr << (24 - 8*(str_index % 4)));
                        if(str_index + 1 == str_len) {
                            if((str_index + 1) % 4 > 0) {
                                M = M | (1 << (23 - 8*(str_index % 4)));
                            }
                        }
                    }

                    str_index++;
                }
            }
            else if(str_index == str_len) {
                M = 1 << 31;
                str_index++;
            }

            Ms[i] = M;
        }
    }

    close(fd);
}

void generate_Ks(unsigned int *Ks) {
    for(int i = 0; i < 64; i++) {
        Ks[i] = (unsigned int) (fabs(sin(i + 1)) * pow(2, 32));
    }
}

unsigned int rotate(unsigned int num, int shift) {
    return (num << shift) | (num >> (32 - shift));
}

unsigned int F(unsigned int B, unsigned int C, unsigned int D) {
    return (B & C) | ((~B) & D);
}

unsigned int G(unsigned int B, unsigned int C, unsigned int D) {
    return (B & D) | (C & (~D));
}

unsigned int H(unsigned int B, unsigned int C, unsigned int D) {
    return B ^ C ^ D;
}

unsigned int I(unsigned int B, unsigned int C, unsigned int D) {
    return C ^ (B | (~D));
}